docs/source/python/dataset.rst

```diff
@@ -456,20 +456,163 @@ is materialized as columns when reading the data and can be used for filtering:
     dataset.to_table().to_pandas()
     dataset.to_table(filter=ds.field('year') == 2019).to_pandas()
 
+Another benefit of manually scheduling the files is that the order of the files
+controls the order of the data.  When performing an ordered read (or a read to
+a table) then the rows returned will match the order of the files given.  This
+only applies when the dataset is constructed with a list of files.  There
+are no order guarantees given when the files are instead discovered by scanning
```

Comment:
We do sort by path: 
https://github.com/apache/arrow/blob/91f261fa9a7841fd914c5ed1d8e747fb4e510a5b/cpp/src/arrow/dataset/discovery.cc#L203-L204

Though it states that's only for unit tests. However I would personally be really surprised if data were truly returned in random order and as we've had this behavior in a few releases now I think users would also be surprised. Maybe we should instead guarantee lexicographic sort on paths?

Reply:
Hmmm, I just dropped the sentence about scanning.  If someone wants clarification we can discuss it at that time.  @bkietz any opinion?  I recall we spoke about this before.

Reply:
I also suppose we don't have to solve this here; we can get the docs in and improve this part later.

Reply:
We don't guarantee order for selectors because ARROW-8163 (asynchronous fragment discovery) might not guarantee order. Lexicographic sorting *could* be maintained for synchronous discovery from a selector, but in general we'd want to push a fragment into scan as soon as it's yielded by `FileSystem::GetFileInfoGenerator`

Reply:
Ah thanks. Sorry for the churn here @westonpace - we should keep stating that there's no guarantee then.